metagen.1
.\" Contact pythonhead@gentoo.org to correct errors or omissions. 
.TH man 1 "22 August 2004" "1.0" "metagen man page"
.SH NAME
.B metagen
\- generate metadata.xml for ebuilds
.SH SYNOPSIS
.B metagen
[options]
.SH DESCRIPTION
.B metagen 
will create a valid metadata.xml file in the current directory. metadata.xml requires a herd tag (-H). If you only specify a package maintainer's email address (-e) "no-herd" will be inserted in the herd tag automatically.

If you need multiple elements, such as two herds, use a comma to
separate them. See EXAMPLES.

.SH OPTIONS
.\" metagen [OPTIONS]
  -H  herd
      Name of herd.

  -e  email-address
      Package maintainer's email address

  -n  maintainer-name
      Package maintainer's name (used with -e option)

  -m 
      Uses ECHANGELOG_USER variable. Can be used instead of -e and -n

  -d  description
      Description of maintainership (used with -e option)

  -l  long-description
      Long description of package.

  -o  output-file
      Write to <output-file> instead of ./metadata.xml

  -f
      Force overwrite of existing metadata

  -v 
      Write to stdout as well as disk (default)

  -q
      Don't write to stdout

  -Q
      Don't write file to disk 

  -h, --help  show this help message and exit

.SH EXAMPLES
.B metagen -H python

Creates metadata.xml in current directory with python as the herd


.B metagen -H python,wxwidgets \
           -l 'This package does a little of this and some o dat.'

Creates metadata in current directory with two herds and long description


.B metagen -e pythonhead@gentoo.org \
           -n 'Joe Blow' \
           -d 'The voices in my head told me to maintain this package' \
           -Q

herd='no-herd', maintainer email, maintainer name, description of maintainership, to stdout only


.B metagen -H net-p2p,python -e jo@gentoo.org,flo@gentoo.org \
           -n 'Jo Blo','Flo Blo' \
           -d 'I am maintaining this because foo','I am maintaining this because bar' \
           -l 'This package is yadda yadda yadda, spam and eggs' \
           -Q

2 herds, 2 maintainers with names and maintainer descriptions for each maintainer,
only write to stdout


.SH FILES
.P 
.I /usr/bin/metagen
.I /usr/share/man/man1/metagen.1.gz
.SH BUGS
No known bugs at this time. 
.SH AUTHOR
.nf
Rob 'pythonhead' Cakebread (pythonhead@gentoo.org)
.fi
.SH HISTORY
2004 \- Initial revision


docs/ChangeLog
* metagen-0.4 (04 Jan 2005)

  04 Jan 04 2005; Rob Cakebread <pythonhead@gentoo.org>:
  Major refactoring. Created separate class metagenerator.py
  Created rudimentary testing framework.



metagen.py
#!/usr/bin/python


"""

NAME     - metagen

SYNOPSIS - Adds metadata.xml to current directory

AUTHOR   - Rob Cakebread <pythonhead@gentoo.org>

USE      - metagen --help

EXAMPLES - man metagen

"""

import sys
import re
import os
from optparse import OptionParser
from commands import getstatusoutput

from portage.output import red, blue

import metagenerator


def parse_echangelog_variable(name, email):
    """Extract developer name and email from ECHANGELOG_USER variable"""
    try:
        e = os.environ["ECHANGELOG_USER"]
    except KeyError:
        print red("!!! Environmental variable ECHANGELOG_USER not set.")
        print red("!!! Set ECHANGELOG_USER or use -e and -n")
        sys.exit(1) 
    try:
        my_email = e[e.find("<") +1:e.find(">")]
    except:
        print red("!!! ECHANGELOG_USER not set properly")
        sys.exit(1) 
    try:
        my_name = e[0:e.find("<")-1]
    except:
        print red("!!! ECHANGELOG_USER not set properly")
        sys.exit(1) 
    if email:
        email = "%s,%s" % (my_email, email)
    else:
        email = my_email
    if name:
        name = "%s,%s" % (my_name, name)
    else:
        name = my_name
    return name, email

def generate_xml(options):
    """Returns metadata.xml text"""

    metadata = metagenerator.MyMetadata()

    if options.herd:
        herds = options.herd.split(",")
    else:
        herds = ["no-herd"]
    metadata.set_herd(herds)

    if options.echangelog:
        (options.name, options.email) = \
            parse_echangelog_variable(options.name, options.email)

    if options.email:
        names, descs = [], []
        if options.name:
            names = options.name.split(",")
        if options.desc:
            descs = options.desc.split(",")
        metadata.set_maintainer(options.email.split(","),
                                names,
                                descs
                                )

    if options.long:
        metadata.set_longdescription(options.long)

    return "%s" % metadata

def validate_xml(my_xml):
    """Test for valid XML"""
    #TODO validate against DTD
    #This just makes sure its valid XML of some sort.
    #Probably not necessary since repoma validates against DTD?
    re_escape_quotes = re.compile('"')
    s = re_escape_quotes.sub('\\"', my_xml)
    cmd = "echo \"%s\" | xmllint --valid - 2>&1 > /dev/null" % s
    return getstatusoutput(cmd)[0]


if __name__ == '__main__':
    optParser = OptionParser()
    optParser.add_option("-H", action="store", dest="herd", type="string",
                         help="Name of herd. If not specified, " +
                         "'no-herd' will be inserted. " +
                         "This requires either the -e or -m option.")

    optParser.add_option("-e", action="store", dest="email", type="string",
                         help="Maintainer's email address")

    optParser.add_option("-n", action="store", dest="name", type="string",
                         help="Maintainer's name")

    optParser.add_option("-m", action="store_true", dest="echangelog", 
                         default=False,
                         help="Use name and email address from ECHANGELOG_USER "+
                         "environmental variable. "+
                         "This is a shortcut for -e <email> -n <name>")

    optParser.add_option("-d", action="store", dest="desc", type="string",
                         help="Description of maintainership")

    optParser.add_option("-l", action="store", dest="long", type="string",
                         help="Long description of package.")

    optParser.add_option("-o", action="store", dest="output", type="string",
                         help="Specify location of output file.")

    optParser.add_option("-f", action="store_true", dest="force", default=False,
                         help="Force overwrite of existing metadata.")

    optParser.add_option("-v", action="store_true", dest="verbose", default=True,
                         help="Verbose. Output of file to stdout. (default)")

    optParser.add_option("-q", action="store_false", dest="verbose",
                         help="Squelch output of file to stdout.")

    optParser.add_option("-Q", action="store_true", dest="no_write",
                         default=False,
                         help="Do not write file to disk.")

    (options, remainingArgs) = optParser.parse_args()

    if len(sys.argv) == 1:
        optParser.print_help()
        sys.exit(1)

    if options.desc or options.name:
        if not options.email and not options.echangelog:
            print red("!!! No maintainer's email address specified.")
            print red("!!! Options -d and -n are only valid with -e or -m")
            sys.exit(1)
 
    if options.herd == "no-herd" and not options.email and not options.echangelog:
        print red("!!! You must specify a maintainer if you have no-herd.")
        sys.exit(1)

    if not options.herd and not options.email and not options.echangelog:
        print red("!!! You must specify at least a herd (-H) " +
                  "or maintainer's email address (-e)\n")
        sys.exit(1)

    txt = generate_xml(options)

    error_status = validate_xml(txt)
    if error_status < 0:
        print red("!!! Error - Invalid XML")
        print red("!!! Please report this bug with the options you used and the output:")
        print error_status
        print txt
        sys.exit(1)

    if options.verbose:
        print "\n%s" % txt

    out_file = "./metadata.xml"
    if options.output:
        out_file = options.output
    if not options.no_write and os.path.exists(out_file):
        if not options.force:
            print red("!!! File %s exists." % out_file)
            print red("!!! Use -f to force overwrite.")
            sys.exit(1)
    if not options.no_write:
        open("%s" % out_file, "w").writelines(txt)
        print blue("%s written") % out_file

